core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1355,51 +1363,45 @@ public void delete(AlluxioURI path, DeleteOptions options) throws IOException,
       }
       // Possible ufs sync.
       syncMetadata(journalContext, inodePath, lockingScheme,
-          options.isRecursive() ? DescendantType.ALL : DescendantType.ONE);
+          options.isRecursive() ? DescendantType.ALL : DescendantType.ONE, blockDeletionContext);
       if (!inodePath.fullPathExists()) {
         throw new FileDoesNotExistException(ExceptionMessage.PATH_DOES_NOT_EXIST.getMessage(path));
       }
 
-      deletedInodes = deleteAndJournal(inodePath, options, journalContext);
+      deleteAndJournal(inodePath, options, journalContext, blockDeletionContext);
       auditContext.setSucceeded(true);
 
-      if (!options.isAlluxioOnly()) {
-        for (Inode<?> inode : deletedInodes) {
-          if (inode.isFile() && inode.isPersisted()) {
-            for (long blockId : ((InodeFile) inode).getBlockIds()) {
-              mUfsBlockLocationCache.invalidate(blockId);
-            }
-          }
-        }
+      for (long blockId : blockDeletionContext.getBlocks()) {
+        mUfsBlockLocationCache.invalidate(blockId);
```

Comment:
should this go to the block delete context close?

Reply:
Does this work correctly with a `noop` context, ie. are there cases where we would use a `noop` context but still want to invalid the cache?

Reply:
Actually, should this move down to `deleteInternal`, because ufs sync will also call delete internal, and any delete should invalidate the block location cache?

Reply:
Moved to the block delete context close.

The noop context is only used during journal replay, where we don't talk to the UFS and we don't need to invalidate the block location cache

With the move to the block deletion context, UFS sync will now invalidate the cache

Reply:
done